File: algorithms/dp/burst_balloons.py
# https://leetcode.com/problems/burst-balloons/
# https://leetcode.com/problems/burst-balloons/discuss/76228/Share-some-analysis-and-explanations

# time complexity:

class Solution:
    def maxCoins(self, nums: list[int]) -> int:
        nums, n, N = [1] + nums + [1], len(nums), len(nums) + 2
        dp = [[0] * N for _ in range(N - 1)]
        for dist in range(2, N):
            for i in range(N - dist):
                k = i + dist
                dp[i][k] = max(dp[i][j] + nums[i] * nums[j] * nums[k] + dp[j][k]
                               for j in range(i + 1, k))
        return dp[0][n + 1]

File: algorithms/dp/burst_balloons.rs
// https://leetcode.com/problems/burst-balloons/
// https://leetcode.com/problems/burst-balloons/discuss/76228/Share-some-analysis-and-explanations

// time complexity:

use std::cmp::max;

impl Solution {
    pub fn max_coins(nums: Vec<i32>) -> i32 {
        let mut nums = nums;
        nums.insert(0, 1);
        nums.push(1);
        let len = nums.len();

        let mut dp = vec![vec![0; len]; len - 1];
        for dist in 2..len {
            for i in 0..len - dist {
                let j = i + dist;
                for k in i + 1..j {
                    dp[i][j] = max(dp[i][j], dp[i][k] + nums[i] * nums[k] * nums[j] + dp[k][j]);
                }
            }
        }
        return dp[0][len - 1];
    }
}